sqlglot/jsonpath.py

```diff
@@ -0,0 +1,210 @@
+from __future__ import annotations
+
+import typing as t
+
+from sqlglot.errors import ParseError
+from sqlglot.expressions import SAFE_IDENTIFIER_RE
+from sqlglot.tokens import Token, Tokenizer, TokenType
+
+if t.TYPE_CHECKING:
+    from sqlglot._typing import Lit
+
+
+class JSONPathTokenizer(Tokenizer):
+    SINGLE_TOKENS = {
+        "(": TokenType.L_PAREN,
+        ")": TokenType.R_PAREN,
+        "[": TokenType.L_BRACKET,
+        "]": TokenType.R_BRACKET,
+        ":": TokenType.COLON,
+        ",": TokenType.COMMA,
+        "-": TokenType.DASH,
+        ".": TokenType.DOT,
+        "?": TokenType.PLACEHOLDER,
+        "@": TokenType.PARAMETER,
+        "'": TokenType.QUOTE,
+        '"': TokenType.QUOTE,
+        "$": TokenType.DOLLAR,
+        "*": TokenType.STAR,
+    }
+
+    KEYWORDS = {
+        "..": TokenType.DOT,
+    }
+
+    IDENTIFIER_ESCAPES = ["\\"]
+    STRING_ESCAPES = ["\\"]
+
+
+JSONPathNode = t.Dict[str, t.Any]
+
+
+def _node(kind: str, value: t.Any = None, **kwargs: t.Any) -> JSONPathNode:
+    node = {"kind": kind, **kwargs}
+
+    if value is not None:
+        node["value"] = value
+
+    return node
+
+
+def parse(path: str) -> t.List[JSONPathNode]:
+    """Takes in a JSONPath string and converts into a list of nodes."""
+    tokens = JSONPathTokenizer().tokenize(path)
+
+    i = 0
+
+    def _curr() -> t.Optional[TokenType]:
+        return tokens[i].token_type if i < len(tokens) else None
+
+    def _prev() -> Token:
+        return tokens[i - 1]
+
+    def _advance() -> Token:
+        nonlocal i
+        i += 1
+        return _prev()
+
+    def _error(msg: str) -> str:
+        return f"{msg} at index {i}: {path}"
+
+    @t.overload
+    def _match(token_type: TokenType, raise_unmatched: Lit[True] = True) -> Token:
+        pass
+
+    @t.overload
+    def _match(token_type: TokenType, raise_unmatched: Lit[False] = False) -> t.Optional[Token]:
+        pass
+
+    def _match(token_type, raise_unmatched=False):
+        if _curr() == token_type:
+            return _advance()
+        if raise_unmatched:
+            raise ParseError(_error(f"Expected {token_type}"))
+        return None
+
+    def _parse_literal() -> t.Any:
+        token = _match(TokenType.STRING) or _match(TokenType.IDENTIFIER)
+        if token:
+            return token.text
+        if _match(TokenType.STAR):
+            return _node("wildcard")
+        if _match(TokenType.PLACEHOLDER):
+            start = i
+
+            while True:
+                if _match(TokenType.L_BRACKET):
+                    _parse_bracket()  # nested call which we can throw away
+                if _curr() in (TokenType.R_BRACKET, None):
+                    break
```

Comment:
If I understand correctly, this allows us to successfully parse `[?(...)`. Do we wanna add another branch here that raises if `_curr()` returns `None` for better error handling?

Reply:
Nvm I just saw https://github.com/tobymao/sqlglot/pull/2867/files#diff-c779839b81bb6254e97ad8a6e17e9efbb09031d113a4942040fdd2cc3dcd9275R145, I think it's handled there.